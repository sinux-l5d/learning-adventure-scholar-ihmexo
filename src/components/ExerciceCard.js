import * as React from 'react';
import Box from '@mui/material/Box';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';
const bull = (
  <Box
    component="span"
    sx={{ display: 'inline-block', mx: '2px', transform: 'scale(0.8)' }}
  >
    •
  </Box>
);

function ExerciceCard({data}) {
  return (
    <Card sx={{ minWidth: 275 }}>
      <CardContent>
        <Typography variant="h5"  gutterBottom>
        {bull}{data.nom}
        </Typography>
        <Typography  color="text.secondary" component="div">
          {bull}Themes:
          {data.theme
            .map((element) => {
                return(<ul>
                    <li>{element}</li>
                </ul>)
            })
          }
        </Typography>
        <Typography sx={{ mb: 1.5 }} color="text.secondary">
        {bull}langage: {data.langage}
        </Typography>
        <Typography variant="body2">
        {bull}difficulté: {data.difficulte}
        </Typography>
      </CardContent>
      <CardActions>
        <Button size="small">Learn More</Button>
      </CardActions>
    </Card>
  );
}
export default ExerciceCard;