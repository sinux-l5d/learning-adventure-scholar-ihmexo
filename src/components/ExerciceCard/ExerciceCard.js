import * as React from "react";
import Box from "@mui/material/Box";
import Card from "@mui/material/Card";
import CardActions from "@mui/material/CardActions";
import CardContent from "@mui/material/CardContent";
import Button from "@mui/material/Button";
import { Typography, ClickAwayListener } from "@mui/material";
import PropTypes, { element } from "prop-types";
import ExerciceInfo from "@components/ExerciceCard/ExerciceInfo/ExerciceInfo";

const bull = (
  <Box component="span" sx={{ display: "inline-block", mx: "2px" }}>
    •
  </Box>
);

/*
 * Carte permettant d'afficher les donnees d'un exercice
 */
function ExerciceCard({ data }) {
  const [openInfo, setOpenInfo] = React.useState(false);
  const fermerInfo = () => {
    setOpenInfo(false);
  };
  const alternerInfo = () => {
    setOpenInfo(!openInfo);
  };

  return (
    <Card sx={{ minWidth: 275, maxWidth: 300 }}>
      <CardContent>
        <Typography variant="h5" gutterBottom>
          {data.nom}
        </Typography>
        <div id={data._id + "-Theme"}>
          {bull}Themes:
          <ul>
            {data.themes.map((element) => {
              return <li key={element}>{element}</li>;
            })}
          </ul>
        </div>
        <Typography sx={{ mb: 1.5 }}>
          {bull}langage: {data.langage}
        </Typography>
        <Typography variant="body2" data-testid="difficulte">
          {bull}difficulté: {data.difficulte}
        </Typography>
      </CardContent>
      <CardActions>
        <ClickAwayListener onClickAway={fermerInfo}>
          <Box>
            <Button onClick={alternerInfo} data-testid="button-info">
              Afficher tous les champs
            </Button>
            {openInfo ? <ExerciceInfo data={data} /> : null}
          </Box>
        </ClickAwayListener>
      </CardActions>
    </Card>
  );
}

/*
 * Types des props passer en parametre au compoment
 * Attention: mettre les bons types sur les props
 */
ExerciceCard.propTypes = {
  data: PropTypes.shape({
    nom: PropTypes.string,
    themes: PropTypes.arrayOf(PropTypes.string),
    langage: PropTypes.string,
    difficulte: PropTypes.number,
    aides: PropTypes.arrayOf(PropTypes.string),
    template: PropTypes.string,
    auteurs: PropTypes.arrayOf(PropTypes.string),
    enonce: PropTypes.string,
    dataset: PropTypes.arrayOf(PropTypes.string),
    _id: PropTypes.string,
  }),
};

export default ExerciceCard;
