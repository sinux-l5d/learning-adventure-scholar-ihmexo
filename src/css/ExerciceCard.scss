.backdrop-exercices {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    background-color: rgb(69, 143, 60);
    border-color: red;
    border-width: 3px;
}
.backdrop-exercice-list{
    display: flex;
    flex-direction: row;
    border-color: red;
    border-width: 3px;
}