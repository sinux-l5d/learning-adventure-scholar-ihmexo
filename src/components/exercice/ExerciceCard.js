import * as React from "react";
import Box from "@mui/material/Box";
import Card from "@mui/material/Card";
import CardActions from "@mui/material/CardActions";
import CardContent from "@mui/material/CardContent";
import Button from "@mui/material/Button";
import { Typography, ClickAwayListener } from "@mui/material";
import PropTypes, { element } from "prop-types";
import Backdrop from "@mui/material/Backdrop";
import "../../css/ExerciceCard.scss";
import ExerciceInfo from "./ExerciceInfo";

const bull = (
  <Box component="span" sx={{ display: "inline-block", mx: "2px" }}>
    •
  </Box>
);

/*
 * Carte permettant d'afficher les donnees d'un exercice
 */
function ExerciceCard({ data }) {
  const [openInfo, setOpenInfo] = React.useState(false);
  const fermerInfo = () => {
    setOpenInfo(false);
  };
  const alternerInfo = () => {
    setOpenInfo(!openInfo);
  };

  return (
    <Card sx={{ minWidth: 275 }}>
      <CardContent>
        <Typography variant="h5" gutterBottom>
          {bull}
          {data.nom}
        </Typography>
        <div id={data._id + "-Theme"}>
          {bull}Themes:
          <ul>
            {data.theme.map((element) => {
              return <li key={element}>{element}</li>;
            })}
          </ul>
        </div>
        <Typography sx={{ mb: 1.5 }} color="text.secondary">
          {bull}langage: {data.langage}
        </Typography>
        <Typography variant="body2">
          {bull}difficulté: {data.difficulte}
        </Typography>
      </CardContent>
      <CardActions>
        {/* <Backdrop open={openInfo} onClick={fermerInfo} sx={{ zIndex: 10 }}> */}
        <ClickAwayListener onClickAway={fermerInfo}>
          <Box>
            <Button onClick={alternerInfo}>Afficher tous les champs</Button>
            {/*---------------------------------------------- Passer ca dans un nouveau components --------------------------------------- */}
            {openInfo ? <ExerciceInfo data={data} /> : null}
            {/*---------------------------------------------- Passer ca dans un nouveau components --------------------------------------- */}
            {/* </Backdrop> */}
          </Box>
        </ClickAwayListener>
      </CardActions>
    </Card>
  );
}

/*
 * Types des props passer en parametre au compoment
 * Attention: mettre les bons types sur les props
 */
ExerciceCard.propTypes = {
  data: PropTypes.shape({
    nom: PropTypes.string,
    theme: PropTypes.arrayOf(PropTypes.string),
    langage: PropTypes.string,
    difficulte: PropTypes.number,
    aides: PropTypes.arrayOf(PropTypes.string),
    template: PropTypes.string,
    auteurs: PropTypes.arrayOf(PropTypes.string),
    enonce: PropTypes.string,
    dataset: PropTypes.arrayOf(PropTypes.string),
    _id: PropTypes.string,
  }),
};

export default ExerciceCard;
