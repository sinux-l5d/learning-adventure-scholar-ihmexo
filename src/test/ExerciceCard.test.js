import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

import ExerciceCard from "../components/exercice/ExerciceCard";
import MockedExerciceInfo from "../components/exercice/ExerciceInfo";

const data = {
  _id: "1",
  nom: "La suite de vanErk",
  difficulte: 8,
  theme: ["suites", "fichiers", "mathématiques"],
  langage: "c",
  template:
    '#include <stdlib.h>\n#include <stdio.h>\n\nvoid vanErk(int K_MAX) \n{\n    /*K_MAX = what is the maximum K allowed to appear in the sequence*/\n    FILE *fp;\n       fp = fopen("./outputVE.txt", "w+");\n    \n}',
  correction:
    '#include <stdlib.h>\n#include <stdio.h>\n\nvoid vanErk(int K_MAX) \n{\n    /*K_MAX = what is the maximum K allowed to appear in the sequence*/\n    FILE *fp;\n       fp = fopen("./outputVE.txt", "w+");\n    \n    int* lastSeen = (int*)calloc(K_MAX,sizeof(int));\n    int lastNum, i, tmpI = 0;\n    while(lastNum < K_MAX)\n    {\n        i++;\n        // Step 1 : print the last number\n        fprintf(fp, "%d, ", lastNum);\n        // Step 2 : find when that last number was printed and throw the diff for the last num\n        tmpI = lastNum;\n        lastNum=!lastSeen[lastNum]?0:i-lastSeen[lastNum];\n        // Step 3 : reload the last time i saw the last last num \n        lastSeen[tmpI]=i;\n    }\n    fprintf(fp, "%d, ", lastNum);\n    free(lastSeen);\n    fclose(fp);\n}',
  aides: [
    'penser à utiliser un fichier sous le nom "outputVE.txt" en mode écriture',
  ],
  auteurs: ["Le Mathématicien refoulé du groupe", "Jacques-Olivier Lachaud"],
  enonce:
    'Ecrivez un programme qui imprime dans un fichier la séquence de van erk (voir le lien : https://www.youtube.com/watch?v=etMJxB-igrc&pp=ugMICgJmchABGAE%3D). Le fichier pour K_MAX = 5 doit contenir "0, 0, 1, 0, 2, "',
  dataset: [],
  __v: 0,
};

// On mock le composent générer dans ExerciceCard pour garder les tests independant
jest.mock("../components/exercice/ExerciceInfo", () => {
  // eslint-disable-next-line react/prop-types
  return function DummyExerciceInfo({ data }) {
    // eslint-disable-next-line react/prop-types
    return <div data-testid="mocked-info">{data.template}</div>;
  };
});

let container = null;
beforeEach(() => {
  // met en place un élément DOM comme cible de rendu
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // nettoie en sortie de test
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("ExerciceCard : s’affiche avec toutes les infos", () => {
  // On génère le component a tester
  act(() => {
    render(<ExerciceCard data={data} />, container);
  });

  expect(container.textContent).toContain("suites");
  expect(container.textContent).toContain("La suite de vanErk");
  expect(container.textContent).toContain("langage: c");

  // On peut ajouter des data-testid aux balise pour effectuer les tests
  const difficulte = document.querySelector("[data-testid=difficulte]");
  expect(difficulte.innerHTML).toContain("difficulté:");

  // On récupère le boutton
  const button = document.querySelector("[data-testid=button-info]");
  //  On click sur le boutton
  act(() => {
    button.dispatchEvent(new MouseEvent("click", { bubbles: true }));
  });

  // On vérifie que le click à bien afficher le component et passant les bons props
  expect(
    container.querySelector("[data-testid='mocked-info']").textContent
  ).toContain(data.template);
});
