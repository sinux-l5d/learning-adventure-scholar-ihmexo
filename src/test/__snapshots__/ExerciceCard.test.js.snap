// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card pour afficher un exo 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-pwngrv-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <h5
      className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
    >
      <span
        className="MuiBox-root css-1ie7f9j"
      >
        •
      </span>
      La suite de vanErk
    </h5>
    <div
      id="1-Theme"
    >
      <span
        className="MuiBox-root css-1ie7f9j"
      >
        •
      </span>
      Themes:
      <ul>
        <li>
          suites
        </li>
        <li>
          fichiers
        </li>
        <li>
          mathématiques
        </li>
      </ul>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body1 css-9eg9l8-MuiTypography-root"
    >
      <span
        className="MuiBox-root css-1ie7f9j"
      >
        •
      </span>
      langage: 
      c
    </p>
    <p
      className="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
    >
      <span
        className="MuiBox-root css-1ie7f9j"
      >
        •
      </span>
      difficulté: 
      8
    </p>
  </div>
  <div
    className="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
  >
    <div
      className="MuiBox-root css-0"
      onClick={[Function]}
      onTouchEnd={[Function]}
    >
      <button
        className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Afficher tous les champs
      </button>
    </div>
  </div>
</div>
`;
